src/lib/comms/sol-socket-impl-zephyr.c

```diff
@@ -0,0 +1,383 @@
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_context *ctx;
+    struct net_tuple *tuple;
+    struct net_buf *netbuf;
+    uint8_t *ptr;
+
+    ctx = socket->unicast_context;
+    SOL_NULL_CHECK(ctx, -ENOTCONN);
+
+    tuple = net_context_get_tuple(ctx);
+
+    netbuf = ip_buf_get_tx(ctx);
+    SOL_NULL_CHECK(netbuf, -ENOMEM);
+
+    /* We set address and port on the packet as if we had received this from
+     * the client we want to send to because net_reply() will revert them
+     * before sending */
+    memcpy(&NET_BUF_IP(netbuf)->srcipaddr, cliaddr->addr.in6, sizeof(cliaddr->addr.in6));
+    memcpy(&NET_BUF_IP(netbuf)->destipaddr, &tuple->local_addr->in6_addr, sizeof(tuple->local_addr->in6_addr));
+    NET_BUF_UDP(netbuf)->srcport = uip_htons(cliaddr->port);
+    NET_BUF_UDP(netbuf)->destport = uip_htons(tuple->local_port);
+    uip_set_udp_conn(netbuf) = net_context_get_udp_connection(ctx)->udp_conn;
+
+    ptr = net_buf_add(netbuf, len);
+    memcpy(ptr, buf, len);
+    ip_buf_appdatalen(netbuf) = len;
+
+    if (net_reply(ctx, netbuf) < 0) {
+        ip_buf_unref(netbuf);
+        return -EIO;
+    }
+
+    return len;
+}
+
+static int
+sol_socket_zephyr_join_group(struct sol_socket *s, int ifindex, const struct sol_network_link_addr *group)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_context *ctx;
+    struct net_tuple *tuple;
+    struct net_addr bindaddr;
+    const char *str;
+    SOL_BUFFER_DECLARE_STATIC(buf, 50);
+
+    if (group->family != SOL_NETWORK_FAMILY_INET6)
+        return -EAFNOSUPPORT;
+
+    tuple = net_context_get_tuple(socket->unicast_context);
+    SOL_NULL_CHECK(tuple, -EINVAL);
+
+    bindaddr.family = sol_network_sol_to_af(group->family);
+    memcpy(&bindaddr.in6_addr, group->addr.in6, sizeof(bindaddr.in6_addr));
+
+    ctx = net_context_get(IPPROTO_UDP, NULL, 0, &bindaddr, tuple->local_port);
+    SOL_NULL_CHECK(ctx, -ENOBUFS);
+
+    sol_ptr_vector_append(&socket->mcast_contexts, ctx);
+
+    str = sol_network_link_addr_to_str(group, &buf);
+    SOL_WRN("Joined group %s", str);
+
+    return 0;
+}
+
+static int
+sol_socket_zephyr_bind(struct sol_socket *s, const struct sol_network_link_addr *addr)
+{
+    struct sol_socket_zephyr *socket = (struct sol_socket_zephyr *)s;
+    struct net_context *ctx;
+    struct net_addr bindaddr;
+    const char *str;
+    SOL_BUFFER_DECLARE_STATIC(buf, 50);
+
+    if (addr->family != SOL_NETWORK_FAMILY_INET6)
+        return -EAFNOSUPPORT;
+
+    if (socket->unicast_context)
+        return -EALREADY;
+
+    bindaddr.family = sol_network_sol_to_af(addr->family);
+    memcpy(&bindaddr.in6_addr, addr->addr.in6, sizeof(bindaddr.in6_addr));
+
+    ctx = net_context_get(IPPROTO_UDP, NULL, 0, &bindaddr, addr->port);
+    SOL_NULL_CHECK(ctx, -ENOBUFS);
+
+    socket->unicast_context = ctx;
+
+    sol_ptr_vector_append(&bound_sockets, socket);
```

Comment:
On Mon, 28 Mar 2016 11:51:11 -0700, Ederson de Souza wrote:
> > +
> > +    if (addr->family != SOL_NETWORK_FAMILY_INET6)
> > +        return -EAFNOSUPPORT;
> > +
> > +    if (socket->unicast_context)
> > +        return -EALREADY;
> > +
> > +    bindaddr.family = sol_network_sol_to_af(addr->family);
> > +    memcpy(&bindaddr.in6_addr, addr->addr.in6, sizeof(bindaddr.in6_addr));
> > +
> > +    ctx = net_context_get(IPPROTO_UDP, NULL, 0, &bindaddr, addr->port);
> > +    SOL_NULL_CHECK(ctx, -ENOBUFS);
> > +
> > +    socket->unicast_context = ctx;
> > +
> > +    sol_ptr_vector_append(&bound_sockets, socket);
> 
> Can't this fail?

Yes. Also, I left so much debug in there...